File: android/src/main/java/expo/modules/bitchat/BitchatModule.kt
package expo.modules.bitchat

import com.bitchat.android.crypto.EncryptionService
import com.bitchat.android.mesh.BluetoothMeshDelegate
import com.bitchat.android.mesh.BluetoothMeshService
import com.bitchat.android.model.BitchatMessage
import com.bitchat.android.model.DeliveryAck
import com.bitchat.android.model.ReadReceipt
import com.bitchat.android.model.DeliveryStatus
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import java.security.MessageDigest
import javax.crypto.spec.SecretKeySpec

// FIX: Correctly implement the delegate interface
class BitchatModule : Module(), BluetoothMeshDelegate {
  private var meshService: BluetoothMeshService? = null
  private val encryptionService: EncryptionService by lazy {
    EncryptionService(appContext.reactContext!!)
  }
  // State provided by JavaScript
  private var nickname: String? = null
  private var favoritePeerIDs = mutableSetOf<String>()
  private var blockedPeerIDs = mutableSetOf<String>()
  private var channelKeys = mutableMapOf<String, SecretKeySpec>()

  override fun definition() = ModuleDefinition {
    Name("Bitchat")

    Events(
        "onMessageReceived",
        "onPeerConnected",
        "onPeerDisconnected",
        "onPeerListUpdated",
        "onDeliveryAck",
        "onReadReceipt",
        "onDeliveryStatusUpdate"
    )

    AsyncFunction("startServices") { promise: Promise ->
      try {
        meshService = BluetoothMeshService(appContext.reactContext!!)
        meshService?.delegate = this@BitchatModule // Correctly assign delegate
        meshService?.startServices()
        promise.resolve(true)
      } catch (e: Exception) {
        promise.reject(CodedException(e))
      }
    }

    AsyncFunction("stopServices") { promise: Promise ->
      try {
        meshService?.stopServices()
        meshService = null
        promise.resolve(true)
      } catch (e: Exception) {
        promise.reject(CodedException(e))
      }
    }

    AsyncFunction("setNickname") { nickname: String ->
      this@BitchatModule.nickname = nickname
    }

    AsyncFunction("toggleFavorite") { peerID: String, isFavorite: Boolean ->
      if (isFavorite) {
        this@BitchatModule.favoritePeerIDs.add(peerID)
      } else {
        this@BitchatModule.favoritePeerIDs.remove(peerID)
      }
    }
    
    AsyncFunction("blockUser") { peerID: String ->
      this@BitchatModule.blockedPeerIDs.add(peerID)
    }

    AsyncFunction("unblockUser") { peerID: String ->
      this@BitchatModule.blockedPeerIDs.remove(peerID)
    }

    AsyncFunction("sendMessage") { content: String, mentions: List<String>, channel: String? ->
      meshService?.sendMessage(content, mentions, channel)
    }

    AsyncFunction("sendPrivateMessage") { content: String, recipientPeerID: String, recipientNickname: String ->
      meshService?.sendPrivateMessage(content, recipientPeerID, recipientNickname)
    }

    AsyncFunction("sendEncryptedChannelMessage") { content: String, mentions: List<String>, channel: String ->
      channelKeys[channel]?.let { key ->
        meshService?.sendEncryptedChannelMessage(content, mentions, channel, key)
      }
    }
    
    AsyncFunction("setChannelPassword") { channel: String, password: String? ->
        if (password != null && password.isNotEmpty()) {
            val key = deriveChannelKey(password, channel)
            this@BitchatModule.channelKeys[channel] = key
        } else {
            this@BitchatModule.channelKeys.remove(channel)
        }
    }

    AsyncFunction("getConnectedPeers") {
      return@AsyncFunction meshService?.getPeerNicknames() ?: emptyMap<String, String>()
    }
    
    AsyncFunction("getPeerRSSI") {
        return@AsyncFunction meshService?.getPeerRSSI() ?: emptyMap<String, Int>()
    }

    AsyncFunction("getDebugStatus") {
      return@AsyncFunction meshService?.getDebugStatus() ?: "Service not initialized"
    }

    AsyncFunction("panicClearAllData") {
        meshService?.stopServices() // A simplified version of panic
    }
  }

  // MARK: - BluetoothMeshDelegate Implementation

  override fun didReceiveMessage(message: BitchatMessage) {
    if (message.senderPeerID != null && blockedPeerIDs.contains(message.senderPeerID)) {
        return
    }
    sendEvent("onMessageReceived", mapOf(
      "id" to message.id,
      "sender" to message.sender,
      "content" to message.content,
      "timestamp" to message.timestamp.time,
      "isPrivate" to message.isPrivate,
      "channel" to message.channel,
      "mentions" to message.mentions,
      "senderPeerID" to message.senderPeerID,
      "isEncrypted" to message.isEncrypted,
      "deliveryStatus" to serializeDeliveryStatus(message.deliveryStatus)
    ))
  }

  override fun didConnectToPeer(peerID: String) {
    sendEvent("onPeerConnected", mapOf("peerID" to peerID, "nickname" to (meshService?.getPeerNicknames()?.get(peerID) ?: peerID)))
  }

  override fun didDisconnectFromPeer(peerID: String) {
    sendEvent("onPeerDisconnected", mapOf("peerID" to peerID, "nickname" to (meshService?.getPeerNicknames()?.get(peerID) ?: peerID)))
  }

  override fun didUpdatePeerList(peers: List<String>) {
    sendEvent("onPeerListUpdated", mapOf("peers" to peers))
  }
  
  override fun didReceiveDeliveryAck(ack: DeliveryAck) {
    sendEvent("onDeliveryAck", mapOf(
        "originalMessageID" to ack.originalMessageID,
        "recipientNickname" to ack.recipientNickname,
        "timestamp" to ack.timestamp.time
    ))
  }

  override fun didReceiveReadReceipt(receipt: ReadReceipt) {
    sendEvent("onReadReceipt", mapOf(
        "originalMessageID" to receipt.originalMessageID,
        "readerNickname" to receipt.readerNickname,
        "timestamp" to receipt.timestamp.time
    ))
  }

  // --- Delegate methods that PROVIDE data to the service ---

  override fun getNickname(): String? {
    return this.nickname
  }

  override fun isFavorite(peerID: String): Boolean {
    return this.favoritePeerIDs.contains(peerID)
  }

  override fun decryptChannelMessage(encryptedContent: ByteArray, channel: String): String? {
    val key = this.channelKeys[channel] ?: return null
    // FIX: Add a call to a new decryption method in your service
    return try {
        encryptionService.decryptWithSymmetricKeyToString(encryptedContent, key)
    } catch (e: Exception) {
        null
    }
  }

  override fun registerPeerPublicKey(peerID: String, publicKeyData: ByteArray) {
    // This is where you could map a peerID to a persistent fingerprint
    // and check against a list of blocked/favorite fingerprints.
  }
  
  override fun didReceiveChannelLeave(channel: String, fromPeer: String) {}

  // MARK: - Helpers

  private fun deriveChannelKey(password: String, channelName: String): SecretKeySpec {
    val salt = channelName.toByteArray(Charsets.UTF_8)
    val combined = password.toByteArray(Charsets.UTF_8) + salt
    val md = MessageDigest.getInstance("SHA-256")
    val keyBytes = md.digest(combined)
    return SecretKeySpec(keyBytes, "AES")
  }
    
  private fun serializeDeliveryStatus(status: DeliveryStatus?): Map<String, Any>? {
    return when (status) {
        is DeliveryStatus.Sending -> mapOf("type" to "sending")
        is DeliveryStatus.Sent -> mapOf("type" to "sent")
        is DeliveryStatus.Delivered -> mapOf("type" to "delivered", "to" to status.to, "at" to status.at.time)
        is DeliveryStatus.Read -> mapOf("type" to "read", "by" to status.by, "at" to status.at.time)
        is DeliveryStatus.Failed -> mapOf("type" to "failed", "reason" to status.reason)
        else -> null
    }
  }
}

File: ios/BitchatModule.swift
import ExpoModulesCore
import CryptoKit

public class BitchatModule: Module, BitchatDelegate {
  private var meshService: BluetoothMeshService?
  
  // State provided by JavaScript and needed by the delegate
  private var nickname: String?
  private var favoritePeerIDs = Set<String>()
  private var blockedPeerIDs = Set<String>()
  private var channelKeys = [String: SymmetricKey]()

  public func definition() -> ModuleDefinition {
    Name("Bitchat")

    Events(
        "onMessageReceived",
        "onPeerConnected",
        "onPeerDisconnected",
        "onPeerListUpdated",
        "onDeliveryAck",
        "onReadReceipt",
        "onDeliveryStatusUpdate"
    )

    AsyncFunction("startServices") { (promise: Promise) in
      meshService = BluetoothMeshService()
      meshService?.delegate = self
      meshService?.startServices()
      promise.resolve(true)
    }

    AsyncFunction("stopServices") { (promise: Promise) in
      meshService?.cleanup()
      meshService = nil
      promise.resolve(true)
    }
    
    // --- State setters from JS ---
    
    AsyncFunction("setNickname") { (nickname: String) in
      self.nickname = nickname
    }

    AsyncFunction("toggleFavorite") { (peerID: String, isFavorite: Bool) in
      if isFavorite {
        self.favoritePeerIDs.insert(peerID)
      } else {
        self.favoritePeerIDs.remove(peerID)
      }
    }
    
    AsyncFunction("blockUser") { (peerID: String) in
        self.blockedPeerIDs.insert(peerID)
    }

    AsyncFunction("unblockUser") { (peerID: String) in
        self.blockedPeerIDs.remove(peerID)
    }

    // --- Message Sending ---

    AsyncFunction("sendMessage") { (content: String, mentions: [String], channel: String?) in
      meshService?.sendMessage(content, mentions: mentions, channel: channel)
    }

    AsyncFunction("sendPrivateMessage") { (content: String, recipientPeerID: String, recipientNickname: String) in
      meshService?.sendPrivateMessage(content, to: recipientPeerID, recipientNickname: recipientNickname)
    }
    
    AsyncFunction("sendEncryptedChannelMessage") { (content: String, mentions: [String], channel: String) in
        guard let key = self.channelKeys[channel] else {
            // Or throw an error to JS
            return
        }
        meshService?.sendEncryptedChannelMessage(content, mentions: mentions, channel: channel, channelKey: key)
    }

    // --- Channel Management ---
    
    AsyncFunction("joinChannel") { (channel: String) -> Bool in
        // Simplified for the bridge; your ViewModel has more complex logic for this
        // In a real app, you might move that logic here or call it from here.
        return true
    }
    
    AsyncFunction("leaveChannel") { (channel: String) in
        meshService?.sendChannelLeaveNotification(channel)
    }
    
    AsyncFunction("setChannelPassword") { (channel: String, password: String?) in
        if let pass = password, !pass.isEmpty {
            let key = deriveChannelKey(from: pass, channelName: channel)
            self.channelKeys[channel] = key
            meshService?.announcePasswordProtectedChannel(channel, isProtected: true, creatorID: meshService?.myPeerID, keyCommitment: computeKeyCommitment(for: key))
        } else {
            self.channelKeys.removeValue(forKey: channel)
            meshService?.announcePasswordProtectedChannel(channel, isProtected: false, creatorID: meshService?.myPeerID, keyCommitment: nil)
        }
    }

    // --- Getters ---
    
    AsyncFunction("getConnectedPeers") {
      return meshService?.getPeerNicknames() ?? [:]
    }
    
    AsyncFunction("getPeerRSSI") {
        return meshService?.getPeerRSSI().mapValues { $0.intValue } ?? [:]
    }
    
    AsyncFunction("getDebugStatus") {
        return "N/A on iOS"
    }

    // --- Safety ---
    
    AsyncFunction("panicClearAllData") {
        meshService?.emergencyDisconnectAll()
        // Here you would also clear Keychain, UserDefaults, etc.
    }
  }

  // MARK: - BitchatDelegate Implementation

  public func didReceiveMessage(_ message: BitchatMessage) {
    // Block check
    if let peerID = message.senderPeerID, self.blockedPeerIDs.contains(peerID) {
        return
    }

    sendEvent("onMessageReceived", [
      "id": message.id,
      "sender": message.sender,
      "content": message.content,
      "timestamp": message.timestamp.timeIntervalSince1970 * 1000,
      "isPrivate": message.isPrivate,
      "channel": message.channel as Any,
      "mentions": message.mentions as Any,
      "senderPeerID": message.senderPeerID as Any,
      "isEncrypted": message.isEncrypted,
      "deliveryStatus": serializeDeliveryStatus(message.deliveryStatus)
    ])
  }

  public func didConnectToPeer(_ peerID: String) {
    sendEvent("onPeerConnected", ["peerID": peerID, "nickname": meshService?.getPeerNicknames()[peerID] ?? peerID])
  }

  public func didDisconnectFromPeer(_ peerID: String) {
    sendEvent("onPeerDisconnected", ["peerID": peerID, "nickname": meshService?.getPeerNicknames()[peerID] ?? peerID])
  }

  public func didUpdatePeerList(_ peers: [String]) {
    sendEvent("onPeerListUpdated", ["peers": peers])
  }

  public func didReceiveDeliveryAck(_ ack: DeliveryAck) {
    sendEvent("onDeliveryAck", [
        "originalMessageID": ack.originalMessageID,
        "recipientNickname": ack.recipientNickname,
        "timestamp": ack.timestamp.timeIntervalSince1970 * 1000
    ])
  }

  public func didReceiveReadReceipt(_ receipt: ReadReceipt) {
    sendEvent("onReadReceipt", [
        "originalMessageID": receipt.originalMessageID,
        "readerNickname": receipt.readerNickname,
        "timestamp": receipt.timestamp.timeIntervalSince1970 * 1000
    ])
  }
  
  public func didUpdateMessageDeliveryStatus(_ messageID: String, status: DeliveryStatus) {
    sendEvent("onDeliveryStatusUpdate", [
        "messageID": messageID,
        "status": serializeDeliveryStatus(status)
    ])
  }

  // --- Delegate methods that PROVIDE data to the service ---

  public func getNickname() -> String? {
    return self.nickname
  }

  public func isFavorite(fingerprint: String) -> Bool {
    // Note: The logic to map peerID to fingerprint would need to be handled
    // For simplicity, we check against our stored peerIDs.
    return self.favoritePeerIDs.contains(fingerprint)
  }
    
  public func decryptChannelMessage(_ encryptedContent: Data, channel: String) -> String? {
    guard let key = self.channelKeys[channel] else { return nil }
    do {
        let sealedBox = try AES.GCM.SealedBox(combined: encryptedContent)
        let decryptedData = try AES.GCM.open(sealedBox, using: key)
        return String(data: decryptedData, encoding: .utf8)
    } catch {
        return nil
    }
  }

  // Default empty implementations for unused delegate methods
  public func didReceiveChannelLeave(_ channel: String, from peerID: String) {}
  public func didReceivePasswordProtectedChannelAnnouncement(_ channel: String, isProtected: Bool, creatorID: String?, keyCommitment: String?) {}
  public func didReceiveChannelRetentionAnnouncement(_ channel: String, enabled: Bool, creatorID: String?) {}

  // MARK: - Helpers
    
  private func deriveChannelKey(from password: String, channelName: String) -> SymmetricKey {
    // This should use the same KDF as your ViewModel
    let salt = channelName.data(using: .utf8)!
    let keyData = HKDF<SHA256>.deriveKey(
        inputKeyMaterial: SymmetricKey(data: password.data(using: .utf8)!),
        salt: salt,
        outputByteCount: 32
    )
    return keyData
  }
    
  private func computeKeyCommitment(for key: SymmetricKey) -> String {
    let keyData = key.withUnsafeBytes { Data($0) }
    let hash = SHA256.hash(data: keyData)
    return hash.compactMap { String(format: "%02x", $0) }.joined()
  }
    
  private func serializeDeliveryStatus(_ status: DeliveryStatus?) -> [String: Any]? {
    guard let status = status else { return nil }
    switch status {
    case .sending:
        return ["type": "sending"]
    case .sent:
        return ["type": "sent"]
    case .delivered(let to, let at):
        return ["type": "delivered", "to": to, "at": at.timeIntervalSince1970 * 1000]
    case .read(let by, let at):
        return ["type": "read", "by": by, "at": at.timeIntervalSince1970 * 1000]
    case .failed(let reason):
        return ["type": "failed", "reason": reason]
    case .partiallyDelivered(let reached, let total):
        // Not used in this implementation, but good to have
        return ["type": "partiallyDelivered", "reached": reached, "total": total]
    }
  }
}